### PR TITLE
[APT-9578] Improve Exception Handling

### PR DESCRIPTION
**Pull request by kabliz**
ArmadilloException needs to support `cause` and `message` in the way `Exception` and most programmers expect them to be used.

Avoids initializing new `Exception()` wherever possible to prevent obfuscating the origin of an error. Never uses `object` for Exceptions, as each carry specific information on their stacktrace.

Exception messages need to be written in plain English, with indication of what component is failing when it fails. And ideally, with a hint on how to fix these errors.

### CODE REVIEW

File: Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt [hunk 1]
```diff
@@ -2,84 +2,103 @@ package com.scribd.armadillo.error
...
+                                message: String = "Unhandled Armadillo Error")
+    : Exception(message, cause) {
     abstract val errorCode: Int
 }
 
 /**
  * Errors in Armadillo
  */
-data class EngineNotInitialized(val reason: String) : ArmadilloException(exception = Exception(reason)) {
+class EngineNotInitialized(message: String)
```
Comment (shreknitin89) [inline Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt:14]: why not make these data classes as well like
```
data class EngineNotInitialized(val reason: String)
    : ArmadilloException(message = "Armadillo Engine is not initialized: $reason") {
    override val errorCode = 100
}
```

File: Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt [hunk 2]
```diff
@@ -2,84 +2,103 @@ package com.scribd.armadillo.error
...
+    : ArmadilloException(message = "Armadillo Engine is not initialized: $message") {
     override val errorCode = 100
 }
 
-object TransportControlsNull : ArmadilloException(Exception()) {
+class TransportControlsNull
+    : ArmadilloException(message = "Internal error, the transport controls are null. Are the controls being initialized?") {
     override val errorCode = 101
 }
```
Comment (shreknitin89) [inline Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt:22]: Similarly keep these as objects as there is no constructor param?
```suggestion
object TransportControlsNull : ArmadilloException(message = "Internal error, the transport controls are null. Are the controls being initialized?") {
    override val errorCode: Int = 101
    private fun readResolve(): Any = this
}
```

File: Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt [hunk 3]
```diff
@@ -2,84 +2,103 @@ package com.scribd.armadillo.error
...
 
-object InvalidPlaybackState : ArmadilloException(Exception()) {
+class InvalidPlaybackState
+    : ArmadilloException(message = "Internal Error, playback state is missing. Has the player been initialized or destroyed?") {
     override val errorCode = 106
 }
 
-data class InvalidRequest(val reason: String) : ArmadilloException(exception = Exception(reason)) {
+class InvalidRequest(message: String)
```
Comment (shreknitin89) [inline Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt:48]: why not keep this as reason and still use `data class`?

File: Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt [hunk 4]
```diff
@@ -2,84 +2,103 @@ package com.scribd.armadillo.error
...
  * Playback Errors
  */
-data class HttpResponseCodeException(val responseCode: Int, val url: String?, val exception: Exception) : ArmadilloException(exception) {
+data class HttpResponseCodeException(val responseCode: Int, val url: String?, override val cause: Exception)
+    : ArmadilloException(cause = cause, message = "HTTP Error $responseCode.") {
     override val errorCode: Int = 200
 }
 
-data class ArmadilloIOException(val exception: Exception) : ArmadilloException(exception) {
+class ArmadilloIOException(cause: Exception, whatActionFailedMessage: String)
```
Comment (shreknitin89) [inline Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt:61]:
```suggestion
data class ArmadilloIOException(val exception: Exception, val reason: String)
```

**Review by shreknitin89 — COMMENTED** [review]
Looks good, but I think we could improve compiler overhead a little bit by using data classes and objects still especially ArmadilloState is being reduced multiple times within playback

File: Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt (hunk 1 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt:14]: adding a data class wouldn't add anything, since `cause` and `message` are both already values in the superclass. 

adding some getters and setters wouldn't really change anything.

File: Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt (hunk 2 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt:22]: I believe using `object` there is causing bugs and interfering with `stacktrace` which can cause annoyances.

File: Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt (hunk 3 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt:48]: There is no need to use `reason` when `message` already exists in the superclass and using `reason` is hiding users from finding `message`. People are expecting `message` and not any weird custom attribute.

**Review by leilaramian — APPROVED** [review]

File: Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt (hunk 1 above)
Comment (shreknitin89) [reply Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt:14]: I'm not really talking about getters and setters. The equality from `equals` would help us resolve the states of immutable objects if the updates are posted to a `conflated` objects like `StateFlows` which internally would use equals method and subside unnecessary emits thus helping us with view redraws or collections is what I think we would benefit from in the long run by using data classes

File: Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt (hunk 2 above)
Comment (shreknitin89) [reply Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt:22]: haa ok..

File: Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt (hunk 3 above)
Comment (shreknitin89) [reply Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt:48]: I'm not suggesting we remove `message` from super class. Just suggesting we use `reason` for the child constructor and construct message using reason .. something like you did with `EngineNotInitialized`
```

data class InvalidRequest(val reason: String) : ArmadilloException(exception = Exception("Armadillo found invalid request: $reason"))

File: Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt (hunk 1 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt:14]: These are `Exception`s. They are for error handling, and they need to conform to norms and expectations set by other `Exception` s.  Armadillo already has a data class,`ErrorAction` which may be what is being confused with here.  `Exceptions` aren't directly for `StateFlows`, these are meant to be raised and caught. And if they *aren't* then I would like to know so I can fix that too.

Furthermore `equals` wouldn't change anything either, because there are no `vals` exclusive in the data class separate from the superclass's equals method.

**Review by shreknitin89 — APPROVED** [review]

File: Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt (hunk 1 above)
Comment (shreknitin89) [reply Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt:14]: ok thanks this is helpful

**Review by rubeus90 — APPROVED** [review]

File: Armadillo/src/main/java/com/scribd/armadillo/ArmadilloPlayerChoreographer.kt [hunk 5]
```diff
@@ -308,7 +308,10 @@ internal class ArmadilloPlayerChoreographer : ArmadilloPlayer {
                 PlaybackState.PLAYING -> controls.pause()
                 PlaybackState.PAUSED -> controls.play()
                 else -> {
-                    stateModifier.dispatch(ErrorAction(UnexpectedException("Unknown state: $playbackState")))
+                    stateModifier.dispatch(ErrorAction(
```
Comment (SeanoAndroid) [inline Armadillo/src/main/java/com/scribd/armadillo/ArmadilloPlayerChoreographer.kt:311]: Somewhat out of scope, but the other `PlaybackState`s such as `STATE_STOPPED` or `STATE_BUFFERING` might not want to throw an exception of the user merely pressed play/pause during those times (I'm assuming it gets prevented downstream, though). Wonder if it's better to just do nothing

File: Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt [hunk 6]
```diff
@@ -89,61 +108,71 @@ object UnableToLoadDownloadInfo : ArmadilloException(Exception()) {
...
     override val errorCode = 305
 }
 
 /**
  * Misc Errors
  */
-data class UnexpectedException(val exception: Exception) : ArmadilloException(exception) {
-    constructor(reason: String) : this(Exception(reason))
-
+class UnexpectedException(cause: Exception, actionThatFailedMessage: String)
```
Comment (SeanoAndroid) [inline Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt:125]: NP: This uses `actionThatFailedMessage` compared to `whatActionFailedMessage` in `ArmadilloIOException`

Might not be a bad idea to keep them synced

**Review by SeanoAndroid — APPROVED** [review]

File: Armadillo/src/main/java/com/scribd/armadillo/ArmadilloPlayerChoreographer.kt (hunk 5 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/ArmadilloPlayerChoreographer.kt:311]: Yeah ill investigate some issues ; i've been ramping up on this project so im hoping that more precise error messaging will help me understand what's going on.

File: Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt (hunk 6 above)
Comment (kabliz) [reply Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt:125]: ok
